modules/js-and-css/optimize-oembeds/hooks.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * Hook callbacks used for oEmbed Optimizer.
+ *
+ * @since n.e.x.t
+ * @package performance-lab
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+$oembed_lazy_load_scripts = false;
+
+/**
+ * Filter the oEmbed HTML.
+ *
+ * Add loading="lazy" to any iframe tags.
+ * Lazy load any script tags.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $html The oEmbed HTML.
+ * @return string
+ */
+function plab_optimize_oembed_html( $html ) {
+	global $oembed_lazy_load_scripts;
+
+	// Find script tags in the oembed using regex.
+	preg_match_all( '/<script\b[^>]*>([\s\S]*?)<\/script>/mi', $html, $matches );
+
+	if ( ! empty( $matches[0][0] ) ) {
+		$found_tag                = $matches[0][0];
+		$oembed_lazy_load_scripts = true;
+
+		// Remove the src attribute from the first script tag, transforming it into data-lazy-src.
+		$script_tag = str_replace( 'src=', 'data-lazy-embed-src=', $matches[0][0] );
+
+		// Replace the script tag with the new one.
+		$html = str_replace( $found_tag, $script_tag, $html );
+	}
+
+	// Find iframes in the oEmbed using a regex.
+	preg_match_all( '/<iframe\b[^>]*>([\s\S]*?)<\/iframe>/mi', $html, $matches );
+
+	// Add loading="lazy" to iframes that don't have a loading attribute.
+	if ( ! empty( $matches[0][0] ) ) {
+		foreach ( $matches[0] as $match ) {
+			if ( false === strpos( $match, 'loading=' ) ) {
+				$iframe_tag = str_replace( '<iframe', '<iframe loading="lazy"', $match );
+				$html       = str_replace( $match, $iframe_tag, $html );
+			}
+		}
+	}
+	return $html;
+}
+add_filter( 'embed_oembed_html', 'plab_optimize_oembed_html', 10 );
+
+/**
+ * Add a script to the footer if there are lazy loaded embeds.
+ * Load the embed's scripts when they approach the viewport using an IntersectionObserver.
+ *
+ * @since n.e.x.t
+ */
+function plab_optimize_oembed_lazy_load_scripts() {
+	global $oembed_lazy_load_scripts;
+
+	if ( ! $oembed_lazy_load_scripts ) {
+		return;
+	}
+	?>
+	<script>
+		(function() {
+			var lazyEmbeds = document.querySelectorAll( 'script[data-lazy-embed-src]' );
```

Comment:
Lazy loading oEmbed scripts with an Intersection Observer feels like a reasonable approach, but I'm somewhat concerned with potential orchestration issues if an oEmbed makes use of multiple scripts; we should ensure that they execute in order when using this approach.

I don't have the opportunity to test things right now myself, but we should also ensure that this works correctly in all major browsers. I seem to recall that some browsers (Firefox, IIRC) don't trigger script execution when changing the `src` in an existing `<script>`, and instead require a new `<script>` tag to be inserted.

Reply:
> Lazy loading oEmbed scripts with an Intersection Observer feels like a reasonable approach, but I'm somewhat concerned with potential orchestration issues if an oEmbed makes use of multiple scripts; we should ensure that they execute in order when using this approach.

So far in my testing of several types, there is a single script tag that loads the oEmbed. We could enforce that, only lazy load if there is a single script. I'd also like to add tests that validate the transforms, we can probably cover all the supported oembed endpoints in the tests.

> I don't have the opportunity to test things right now myself, but we should also ensure that this works correctly in all major browsers. I seem to recall that some browsers (Firefox, IIRC) don't trigger script execution when changing the src in an existing <script>, and instead require a new <script> tag to be inserted.

Ah yes, I was expecting to run into some browser compatibility issues and have not tested yet.  Hopefully we can find a single approach that works across our supported browsers - inserting a new script tag sounds reasonable. I started with this simple src approach for a POC. I will research which approach is best. I also wonder if we can increase the threshold for low bandwidth connections (so the oEmbeds start loading earlier).

Reply:
we are now using a different pattern where a new script tag is created and inserted. I will be sure to test this across recent browsers.

Reply:
we are now using a different pattern where a new script tag is created and inserted. I will be sure to test this across recent browsers.